### PR TITLE
Add SSL/Basic auth support to SR app

### PR DESCRIPTION
**Pull request by rnpridgeon**

### CODE REVIEW

File: trivup/apps/SchemaRegistryApp.py [hunk 1]
```diff
@@ -28,96 +28,115 @@
...
 from trivup.apps.KafkaBrokerApp import KafkaBrokerApp
+from trivup.apps.SslApp import SslApp
 
-import uuid
 import requests
-import subprocess
 
 
-class SchemaRegistryApp (trivup.App):
+sr_image = 'confluentinc/cp-schema-registry'
```
Comment (edenhill) [inline trivup/apps/SchemaRegistryApp.py:36]: Why is this not on the class when the default version is?

File: trivup/apps/SchemaRegistryApp.py [hunk 2]
```diff
@@ -28,96 +28,115 @@
...
+        'host.name': 'localhost',
+        'listeners': 'http://0.0.0.0:8081'
+    }
 
-    def __init__(self, cluster, conf=None, on=None):
+    @staticmethod
+    def _conf_to_env(app_conf):
+        return ["SCHEMA_REGISTRY_{}={}"
+                .format(key.replace(".", "_").upper(), value)
+                for key, value in app_conf.items()]
```
Comment (edenhill) [inline trivup/apps/SchemaRegistryApp.py:56]: is this py2 and py3 compatible?

File: trivup/apps/SchemaRegistryApp.py [hunk 3]
```diff
@@ -28,96 +28,115 @@
...
         Supported conf keys:
            * version - Confluent Platform version to use.
+           * enable_auth - Enable Http basic auth
            * port_base - Low TCP port base to start allocating from (random)
-           * image - docker image to use
-           * conf - schema-registry docker image config strings (NOT USED)
+           * conf - Dictionary of Confluent Schema Registry configs
         """
-        super(SchemaRegistryApp, self).__init__(cluster, conf=conf, on=on)
+        print(conf.get('enable_auth'))
```
Comment (edenhill) [inline trivup/apps/SchemaRegistryApp.py:70]: -

File: trivup/apps/SchemaRegistryApp.py [hunk 4]
```diff
@@ -28,96 +28,115 @@
...
         """
         @param cluster     Current cluster
         @param conf        Configuration dict, see below.
         @param on          Node name to run on
 
         Supported conf keys:
            * version - Confluent Platform version to use.
+           * enable_auth - Enable Http basic auth
            * port_base - Low TCP port base to start allocating from (random)
-           * image - docker image to use
```
Comment (edenhill) [inline trivup/apps/SchemaRegistryApp.py:53]: why remove the ability to specify an alternate image?

File: trivup/apps/SchemaRegistryApp.py [hunk 5]
```diff
@@ -28,96 +28,115 @@
...
+        self.expose_port(self.docker_args, 8081, port)
 
-        if self.conf.get('image', '') == '':
-            self.conf['image'] = self.default_image
+        app_conf = dict(self.default_app_conf, **conf.get('conf', {}))
 
-        self.conf['container_id'] = 'trivup_sr_%s' % str(uuid.uuid4())[0:7]
         kafka = cluster.find_app(KafkaBrokerApp)
-        if kafka is None:
+        if kafka:
```
Comment (edenhill) [inline trivup/apps/SchemaRegistryApp.py:87]: You should avoid if blocks (extra indentation) for the sunny path.
So instead of:
```
if some_err:
   bail out
else:
   the normal case
```

(the else block is unnecessary) 

do what the original code did:
```
if some_err:
   bail out

the normal case
```

File: trivup/apps/SchemaRegistryApp.py [hunk 6]
```diff
@@ -28,96 +28,115 @@
...
-            self.conf.get('listeners'),
-            self.conf.get('image'))
-
-        # Stop through docker
-        self.conf['stop_cmd'] = 'docker stop %s' % \
-                                self.conf.get('container_id')
+        ssl = cluster.find_app(SslApp)
+        if ssl is not None:
+            # Mount SslApp dir, reuse broker trusts/certs
+            self.add_mount(self.docker_args, "/opt/ssl", ssl.root_path())
```
Comment (edenhill) [inline trivup/apps/SchemaRegistryApp.py:96]: /opt may interfer with what is already on the image, use something unique instead, e.g., /trivup/ssl

File: trivup/apps/SchemaRegistryApp.py [hunk 7]
```diff
@@ -28,96 +28,115 @@
...
+            app_conf.update({
+                'listeners': 'https://0.0.0.0:8081',
+                'inter_instance_protocol': 'https',
+                'ssl_keystore_location':
+                    '/opt/ssl/broker{}.keystore.jks'.format(kafka.appid),
+                'ssl_keystore_password': ssl.conf.get('ssl_key_pass'),
+                'ssl_key_password': ssl.conf.get('ssl_key_pass'),
+                'ssl_truststore_location':
+                    '/opt/ssl/broker{}.truststore.jks'.format(kafka.appid),
+                'ssl_truststore_password':  ssl.conf.get('ssl_key_pass')
```
Comment (edenhill) [inline trivup/apps/SchemaRegistryApp.py:107]: two whitespaces. Did you flake8?

File: trivup/apps/SchemaRegistryApp.py [hunk 8]
```diff
@@ -28,96 +28,115 @@
...
+                })
+
+        if bool(self.conf.get('enable_auth', False)):
+            self.conf['userinfo'] = 'ckp_tester:test_secret'
+            app_conf.update({'authentication.method': 'BASIC',
+                             'authentication.realm': 'SchemaRegistry',
+                             'authentication.roles': 'Testers'})
+            app_conf['OPTS'] =\
+                '-Djava.security.auth.login.config=/opt/app/'\
```
Comment (edenhill) [inline trivup/apps/SchemaRegistryApp.py:116]: dito, stay out of /opt

File: trivup/apps/SchemaRegistryApp.py [hunk 9]
```diff
@@ -28,96 +28,115 @@
...
+
+        if bool(self.conf.get('enable_auth', False)):
+            self.conf['userinfo'] = 'ckp_tester:test_secret'
+            self.add_mount(self.docker_args, "/opt/app", self.root_path())
+            self.create_file_from_template('login.properties')
+            self.create_file_from_template('schema-registry.jaas', subst=False)
+
+        self.conf['url'] = '{}://{}@localhost:{}'.format(
+                protocol,
+                self.conf.get('userinfo', ''),
```
Comment (edenhill) [inline trivup/apps/SchemaRegistryApp.py:127]: Is it really ok to have an empty user:pass in a url?

File: trivup/apps/SchemaRegistryApp.py [hunk 10]
```diff
@@ -28,96 +28,115 @@
...
+                port)
+
+        # format configs for use by docker
```
Comment (edenhill) [inline trivup/apps/SchemaRegistryApp.py:130]: docker itself does not care about these envs, maybe 'the schema-registry docker image'

File: trivup/apps/SchemaRegistryApp/login.properties.template [hunk 11]
```diff
@@ -0,0 +1 @@
+${userinfo}, Testers
```
Comment (edenhill) [inline trivup/apps/SchemaRegistryApp/login.properties.template:1]: missing newline

File: trivup/apps/SchemaRegistryApp/schema-registry.jaas.template [hunk 12]
```diff
@@ -0,0 +1,4 @@
+SchemaRegistry {
+    org.eclipse.jetty.jaas.spi.PropertyFileLoginModule required
+    file="/opt/app/login.properties";
+  };
```
Comment (edenhill) [inline trivup/apps/SchemaRegistryApp/schema-registry.jaas.template:4]: missing newline

File: trivup/clusters/KafkaCluster.py [hunk 13]
```diff
@@ -101,6 +102,7 @@ def __init__(self, **kwargs):
         self.cluster = Cluster(
             self.__class__.__name__,
             os.environ.get('TRIVUP_ROOT', 'tmp-%s' % self.__class__.__name__),
+            nodes=['localhost', 'docker'],
```
Comment (edenhill) [inline trivup/clusters/KafkaCluster.py:105]: we only want docker nodes if sr is enabled

File: trivup/clusters/KafkaCluster.py [hunk 14]
```diff
@@ -158,9 +160,12 @@ def __init__(self, **kwargs):
         # Create SchemaRegistry if enabled
         if bool(self.conf.get('with_sr', False)):
             self.sr = SchemaRegistryApp(
-                self.cluster, {'version': self.conf.get('cp_version')})
+                self.cluster, {'version': self.conf.get('cp_version'),
+                               'enable_auth': self.conf.get('with_sr_auth')})
             self.env['SR_URL'] = self.sr.get('url')
 
+        if bool(self.conf.get('with_sr_auth')):
```
Comment (edenhill) [inline trivup/clusters/KafkaCluster.py:167]: The other ones have a default value of False

File: trivup/clusters/KafkaCluster.py [hunk 15]
```diff
@@ -158,9 +160,12 @@ def __init__(self, **kwargs):
...
+            self.env['SR_USERINFO'] = self.sr.get('userinfo')
```
Comment (edenhill) [inline trivup/clusters/KafkaCluster.py:168]: Why is this needed? Isn't the URL enough?

File: trivup/trivup.py [hunk 16]
```diff
@@ -283,6 +284,8 @@ def __init__(self, name):
         self.name = name
         if name == 'localhost':
             self.exec_cmd = ''
+        elif name == 'docker':
+            self.exec_cmd = 'docker'
```
Comment (edenhill) [inline trivup/trivup.py:288]: This is not really the same thing, should probably be `docker run ..` if anything.

File: trivup/trivup.py [hunk 17]
```diff
@@ -642,5 +645,72 @@ def __str__(self):
                                    self.appid, self.state)
 
 
+class DockerApp(App):
```
Comment (edenhill) [inline trivup/trivup.py:648]: no docstring

File: trivup/trivup.py [hunk 18]
```diff
@@ -642,5 +645,72 @@ def __str__(self):
...
+    __slots__ = ['image', 'container_name', 'docker_args', "app_args"]
```
Comment (edenhill) [inline trivup/trivup.py:649]: inconsistent quoting.

what do you need the slots for?

File: trivup/trivup.py [hunk 19]
```diff
@@ -642,5 +645,72 @@ def __str__(self):
...
+
+    def __init__(self, cluster, image, tag, conf):
+        self.image = "{}:{}".format(image, tag)
+        self.conf = conf
+        self.docker_args = []
+        self.app_args = []
+        self.container_name = 'trivup_{}_{}'\
+            .format(image.rsplit("/")[-1], str(uuid4())[0:7])
+
+        super(DockerApp, self).__init__(cluster, conf=conf, on='docker')
```
Comment (edenhill) [inline trivup/trivup.py:659]: These typically go first, right?

File: trivup/trivup.py [hunk 20]
```diff
@@ -642,5 +645,72 @@ def __str__(self):
...
+        return " run -a stdout -a stderr " \
+               "{} --name {} {} {}"\
+                   .format(self._docker_opts_to_list(self.docker_args),
+                           self.container_name,
+                           self._dockerize_app_props(self.app_args),
+                           self.image)
+
+    @staticmethod
+    def _dockerize_app_props(app_conf):
+        return " ".join("-e {}".format(str(prop))
```
Comment (edenhill) [inline trivup/trivup.py:671]: docstrings all over the place, please.

**Review by edenhill — CHANGES_REQUESTED** [review]
Does this work on Py 2 and Py 3, on Linux and OSX?

File: trivup/apps/SchemaRegistryApp.py (hunk 2 above)
Comment (rnpridgeon) [reply trivup/apps/SchemaRegistryApp.py:56]: for i in "2.7.16" "3.7.4"; do pyenv shell $i;  python -c 'import sys; print(sys.version); app_conf = {"a": "b", "c": "d"} ; print(["SCHEMA_REGISTRY_{}={}".format(key.replace(".","_").upper(), value) for key, value in app_conf.items()])'; done

2.7.16 (default, Jul 26 2019, 10:11:46) 
[GCC 4.2.1 Compatible Apple LLVM 10.0.1 (clang-1001.0.46.4)]
['SCHEMA_REGISTRY_A=b', 'SCHEMA_REGISTRY_C=d']

3.7.4 (default, Jul 26 2019, 10:18:18) 
[Clang 10.0.1 (clang-1001.0.46.4)]
['SCHEMA_REGISTRY_A=b', 'SCHEMA_REGISTRY_C=d']
localhost:trivup ryan$

File: trivup/apps/SchemaRegistryApp.py (hunk 4 above)
Comment (rnpridgeon) [reply trivup/apps/SchemaRegistryApp.py:53]: In practice we never pass an alternative image so this seemed simpler. Can revert though

File: trivup/apps/SchemaRegistryApp.py (hunk 9 above)
Comment (rnpridgeon) [reply trivup/apps/SchemaRegistryApp.py:127]: requests allowed it but that may be relying on a bug so I'll make it conditional

File: trivup/trivup.py (hunk 16 above)
Comment (rnpridgeon) [reply trivup/trivup.py:288]: thats fair maybe it just doesn't belong here at all.

File: trivup/trivup.py (hunk 18 above)
Comment (rnpridgeon) [reply trivup/trivup.py:649]: Need is a strong word. I just like that it prevents users from adding attributes arbitrarily.

File: trivup/apps/SchemaRegistryApp.py (hunk 10 above)
Comment (rnpridgeon) [reply trivup/apps/SchemaRegistryApp.py:130]: Sure but docker uses these arguments to configure the application which resides in that image. -e "ENV_VAR" is pretty much the standard way of doing that so I'm not convinced this statement is so terrible.

File: trivup/trivup.py (hunk 18 above)
Comment (edenhill) [reply trivup/trivup.py:649]: I'm not sure I buy into that, it is just more code to maintain for little gain. Remove it.

File: trivup/apps/SchemaRegistryApp.py (hunk 4 above)
Comment (edenhill) [reply trivup/apps/SchemaRegistryApp.py:53]: Please add it back
